File: hashing/0383-ransom-note/java/Solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        Map<Character, Integer> counts = new HashMap<>();

        for (char c : magazine.toCharArray()) {
            counts.merge(c, 1, Integer::sum);
        }

        for (char c : ransomNote.toCharArray()) {
            int remaining = counts.getOrDefault(c, 0);
            if (remaining == 0) return false;
            counts.put(c, remaining - 1);
        }

        return true;
    }
}

File: hashing/0383-ransom-note/cpp/Solution.cpp
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        unordered_map<char, int> counts;

        for (char c : magazine) {
            counts[c]++;
        }

        for (char c : ransomNote) {
            if (counts[c] == 0) return false;
            counts[c]--;
        }

        return true;
    }
};
